File: sort/CountingSort.py
data = [15,2,26,26,26,26,73,48,59,3,95,74,84,15,49,74]

def counting_sort(data):
    max_val = max(data)

    count = [0] * (max_val + 1)

    for num in data:
        count[num] += 1

    for i in range(1, len(count)):
        count[i] += count[i - 1]

    output = [0] * len(data)

    for num in data:
        output[count[num] - 1] = num
        count[num] -= 1

    for i in range(len(data)):
        data[i] = output[i]
    return data

print(counting_sort(data))

File: sort/CountingSort.java
package sort;

public class CountingSort {
    public static void main(String[] args) {
        int[] data = {52,25,6,27,16,27,38,74,36,57,6,8};

        data = counting_sort(data);

        for(int i: data){
            System.out.print(i + "   ");
        }
    }

    public static int[] counting_sort(int[] data){
        int max_val = data[0];                      //set max_val

        for(int num: data){                        //for each(num-data)
            if(num > max_val){                      //if num > max_val
                max_val = num;                      // mv = n
            }
        }
        
        int[] count = new int[max_val + 1];         // count = new int[mv + 1]

        for(int num : data){                        // for each (num-data)
            count[num]++;                           // increment count
        }

        int index = 0;                              //index

        for(int i = 0; i <= max_val; i++){          //for <= max_val
            while(count[i] > 0){                    //while count i > 0
                data[index++] = i;
                count[i]--;
            }
        }
        return data;
    }
}
